frontend/src/metabase/visualizations/lib/settings.js

```diff
@@ -264,3 +265,25 @@ export function updateSettings(
   }
   return newSettings;
 }
+
+// Merge two settings objects together.
+// Settings from the second argument take precedence over the first.
+export function mergeSettings(first: Settings = {}, second: Settings = {}) {
+  // Note: This hardcoded list of all nested settings is potentially fragile,
+  // but both the list of nested settings and the keys used are very stable.
+  const nestedSettings = ["series_settings", "column_settings"];
```

Comment:
I know nothing about the `Settings` object or how often new props are added to them, but it seems like we could avoid hardcoding these property names by iterating over `first` and `second` and collecting all property names with object/array-like values. Would that be possible?

is there any reason we can't just create a list of all object/array-like properties in the two Settings objects on the fly and then iterate over that?

Reply:
I like the idea of checking based on type, but I don't think it will work here. Setting values can be a variety of types. Some examples:

- They're mostly simple types
e.g. `"graph.x_axis.labels_enabled": false`
- They're pretty often arrays
e.g. `"graph.dimensions": ['some column', 'another column']`
- They can also be more deeply nested structures
e.g.
`"table.column_formatting": [{"columns":["TAX","SUBTOTAL","TOTAL"],"type":"range","colors":["white","#509EE3"],"min_type":null,"max_type":"all","min_value":0,"max_value":100,"operator":"=","value":"","color":"#509EE3","highlight_row":false}]`
- Having an object value is less common, but one place is `pivot_table.column_split` that tracks how columns are partitioned and looks like `{rows: [], columns: [], values: []}`.

All of those are normal rather than nested settings. They're all set at the top level of the settings object. Even for array and object values, we wouldn't want to try and merge the values since that could get some funky results. (Though, given the current dashboard UI, I don't think that situation would come up.)

Nested settings repeat the same settings for each "key". Here's a fake example:

```
{
  "some_top_level_key": "value",
  "column_settings": {
    "column identifier 1": {
      "column_specific_setting": "value 1",
    },
    "column identifier 2": {
      "column_specific_setting": "value 2",
    },
  }
}
```

The only two places we use [nested settings](https://github.com/metabase/metabase/blob/master/frontend/src/metabase/visualizations/lib/settings/nested.js) are [column](https://github.com/metabase/metabase/blob/master/frontend/src/metabase/visualizations/lib/settings/column.js) and [series settings](https://github.com/metabase/metabase/blob/master/frontend/src/metabase/visualizations/lib/settings/series.js). Column settings are things like the column title and formatting options. Series settings let you control things like whether to show dots on a line line in a chart or what to label it in the legend.

---

It's definitely possible to avoid hardcoding these. We could either export constants from the two current uses or have `nestedSettings` (which column and series both call) "register" the id. Both of those seemed slightly too complicated[1] given that these have been unchanged for years but might make sense down the line.

[1] Exporting constants isn't complicated, but I was worried about hitting circular import issues without restructuring things.

Reply:
Awesome, thanks for the thorough explanation. Makes sense to me.